### PR TITLE
feat: Add extractors package

### PR DESCRIPTION
**Pull request by sixcolors**
## Summary

This PR introduces a shared `github.com/gofiber/fiber/v3/extractors` package that consolidates value extraction utilities previously duplicated across multiple middleware packages. The package provides a clean, consistent API for extracting values from various sources (headers, cookies, query parameters, form data, URL parameters) with built-in chain/fallback logic, custom extraction support, and comprehensive security considerations.

### New Package: extractors
- **extractors.go**: Core implementation with `Extractor` interface and concrete implementations:
  - `FromAuthHeader` - Extracts from Authorization header with scheme support
  - `FromCookie` - Extracts from HTTP cookies  
  - `FromParam` - Extracts from URL parameters
  - `FromForm` - Extracts from form data
  - `FromHeader` - Extracts from HTTP headers
  - `FromQuery` - Extracts from query parameters
  - `FromCustom` - Custom extraction logic with source awareness
  - `Chain` - Combines multiple extractors with fallback logic
- **extractors_test.go**: Comprehensive test suite with **17 test functions** covering:
  - Basic extraction functionality across all source types
  - Chain/fallback behavior and error propagation
  - Custom extractor support and edge cases
  - RFC-compliant Authorization header parsing
  - Source introspection and metadata validation
  - Security edge cases and whitespace handling

### Documentation Updates
- **extractors/README.md**: Maintainer-focused documentation covering:
  - Package architecture and design decisions
  - Testing strategy and coverage (17 comprehensive tests)
  - Maintenance guidelines and security considerations
  - Source inspection capabilities for security-sensitive operations
- **docs/guide/extractors.md**: User-focused documentation with:
  - Complete usage examples and integration patterns
  - Security warnings and best practices guide
  - Troubleshooting section with common issues
  - Advanced usage patterns including custom extractors
  - Migration guidance for middleware adoption

## Migration Strategy

This PR establishes the foundation for migrating middleware packages to use the shared extractors. Future PRs will update individual middleware packages (session, csrf, keyauth, etc.) to import and use `github.com/gofiber/fiber/v3/extractors` directly instead of maintaining their own extraction logic.

## Benefits

1. **Code Deduplication**: Eliminates ~500+ lines of duplicated extraction code across middleware
2. **Consistency**: Standardized extraction behavior and error handling
3. **Security**: Built-in source inspection and security warnings for different extraction sources
4. **Extensibility**: Easy to add new extraction sources or modify existing behavior
5. **Maintainability**: Single source of truth for extraction logic with comprehensive test coverage

## Testing

- ✅ All **17 test functions** pass with comprehensive coverage
- ✅ Edge cases and error scenarios validated
- ✅ Chain/fallback logic thoroughly tested
- ✅ RFC compliance verified for auth headers
- ✅ Source awareness and metadata validation confirmed

## Security Considerations

The package includes comprehensive security features:
- **Source Awareness**: All extractors track their origin (header, cookie, query, etc.)
- **Security Warnings**: Built-in warnings for potentially insecure sources
- **Input Validation**: Automatic trimming and empty value handling
- **RFC Compliance**: Proper Authorization header parsing
- **Developer Guidance**: Clear documentation on security trade-offs

## Next Steps

After this PR is merged, individual middleware packages will be updated to:
1. Remove their local extractor implementations
2. Import `github.com/gofiber/fiber/v3/extractors`
3. Update usage to call extractors directly
4. Remove any wrapper functions

This migration will be done incrementally across multiple PRs to ensure stability and allow for thorough testing of each middleware package.

### CODE REVIEW

File: extractors/go.mod [hunk 1]
```diff
@@ -0,0 +1,29 @@
+module github.com/gofiber/fiber/v3/extractors
+
+go 1.25.0
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply extractors/go.mod]: @gemini-code-assist you are wrong.

**Comment by sixcolors** [comment]
@coderabbitai review

File: extractors/extractors.go [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+
+	// SourceCustom indicates the value is extracted using a custom extractor function.
+	SourceCustom
+)
+
+// ErrNotFound is returned when the requested value is missing or empty.
+var ErrNotFound = errors.New("value not found")
+
+// Extractor defines a value extraction method with metadata.
+type Extractor struct {
```
Comment (gaby) [inline extractors/extractors.go:67]: Run make betteralign

File: extractors/extractors.go [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+	return Extractor{
+		Extract: func(c fiber.Ctx) (string, error) {
+			authHeader := c.Get(fiber.HeaderAuthorization)
+			if authHeader == "" {
+				return "", ErrNotFound
+			}
+
+			// Check if the header starts with the specified auth scheme
+			if authScheme == "" {
+				return strings.TrimSpace(authHeader), nil
```
Comment (gaby) [inline extractors/extractors.go]: Use utils.Trim

File: extractors/extractors.go [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+//
+// Example:
+//
+//	extractor := FromCookie("session_id")
+//	// Cookie: "session_id=abc123" -> Output: "abc123"
+//	// Missing cookie -> Output: ErrNotFound
+func FromCookie(key string) Extractor {
+	return Extractor{
+		Extract: func(c fiber.Ctx) (string, error) {
+			value := strings.TrimSpace(c.Cookies(key))
```
Comment (gaby) [inline extractors/extractors.go]: Use utils.Trim

File: extractors/extractors.go [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+// Example:
+//
+//	// Route: GET /users/:userId/posts/:postId
+//	userExtractor := FromParam("userId")
+//	postExtractor := FromParam("postId")
+//	// URL: /users/123/posts/456 -> userId: "123", postId: "456"
+func FromParam(param string) Extractor {
+	return Extractor{
+		Extract: func(c fiber.Ctx) (string, error) {
+			value := strings.TrimSpace(c.Params(param))
```
Comment (gaby) [inline extractors/extractors.go]: Use utils.Trim

File: extractors/extractors.go [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+//
+// Example:
+//
+//	extractor := FromForm("username")
+//	// Form data: "username=john_doe&password=secret" -> Output: "john_doe"
+//	// Missing field -> Output: ErrNotFound
+func FromForm(param string) Extractor {
+	return Extractor{
+		Extract: func(c fiber.Ctx) (string, error) {
+			value := strings.TrimSpace(c.FormValue(param))
```
Comment (gaby) [inline extractors/extractors.go]: Use utils.Trim

File: extractors/extractors.go [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+//
+// Example:
+//
+//	extractor := FromHeader("X-API-Key")
+//	// Header: "X-API-Key: abc123" -> Output: "abc123"
+//	// Missing header -> Output: ErrNotFound
+func FromHeader(header string) Extractor {
+	return Extractor{
+		Extract: func(c fiber.Ctx) (string, error) {
+			value := strings.TrimSpace(c.Get(header))
```
Comment (gaby) [inline extractors/extractors.go]: Use utils.Trim

File: extractors/extractors.go [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+//
+// Example:
+//
+//	extractor := FromQuery("token")
+//	// URL: /api/data?token=abc123&format=json -> Output: "abc123"
+//	// URL: /api/data?format=json -> Output: ErrNotFound
+func FromQuery(param string) Extractor {
+	return Extractor{
+		Extract: func(c fiber.Ctx) (string, error) {
+			value := strings.TrimSpace(c.Query(param))
```
Comment (gaby) [inline extractors/extractors.go]: Use utils.Trim

File: extractors/README.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+- **Single Source of Truth**: All extraction logic lives here
+- **Direct Usage**: Middleware imports and uses extractors directly
+- **Security Consistency**: Security warnings and source awareness must be kept in sync across all extractors
+- **Breaking Changes**: Require coordinated updates across dependent packages
+- **Performance**: Shared functions reduce overhead across middleware
+
+## Future Extensions
+
+Potential enhancements:
```
Comment (gaby) [inline extractors/README.md]: Remove this section, instead create an Issue ticket to address them

**Comment by gaby** [comment]
@codex As a security auditor, perform a full RFC compliance check of extractors/extractors.go, also verify the unit-tests are checking for RFC compliance

**Comment by sixcolors** [comment]
@gaby Thanks for the suggestion! I investigated `utils.Trim` and found it would require chained calls `utils.Trim(utils.Trim(value, ' '), '\t')` which introduces ordering dependencies and is slower. The current `strings.Trim(value, " \t")` provides the optimal balance of RFC compliance and performance - it's only ~15-20% slower than `strings.TrimSpace`) but ensures HTTP standard compliance by only trimming SP and HTAB characters.

File: extractors/extractors.go (hunk 8 above)
Comment (ReneWerner87) [reply extractors/extractors.go]: https://github.com/gofiber/fiber/pull/3725#issuecomment-3254965571

File: docs/guide/extractors.md [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+// Avoid: Same source for different middleware
+app.Use(keyauth.New(keyauth.Config{
+    Extractor: extractors.FromCookie("token"), // API auth
+}))
+app.Use(session.New(session.Config{
+    Extractor: extractors.FromCookie("token"), // Session - CONFLICT!
+}))
+```
+
+<parameter name="filePath">/Users/sixcolors/Documents/GitHub/fiber/docs/guide/extractors.md
```
Comment (ReneWerner87) [inline docs/guide/extractors.md]:
```suggestion
```

Comment (sixcolors) [reply docs/guide/extractors.md]: 🤦‍♂️

File: extractors/extractors.go [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+//
+//	// Extract Bearer token
+//	extractor := FromAuthHeader("Bearer")
+//	// Input: "Bearer abc123" -> Output: "abc123"
+//	// Input: "Basic dXNlcjpwYXNz" -> Output: ErrNotFound
+//
+//	// Extract raw header value
+//	extractor := FromAuthHeader("")
+//	// Input: "CustomAuth token123" -> Output: "CustomAuth token123"
+func FromAuthHeader(authScheme string) Extractor {
```
Comment (gaby) [inline extractors/extractors.go:101]: We should copy the one from keyauth:

https://github.com/gofiber/fiber/blob/main/middleware/keyauth/extractors.go#L51

File: extractors/extractors.go [hunk 12]
```diff
@@ -0,0 +1,511 @@
+package extractors
+
+// Package extractors provides shared value extraction utilities for Fiber middleware.
+// This package helps reduce code duplication across middleware packages
+// while ensuring consistent behavior, security practices, and RFC compliance.
+// It can extract string values from various HTTP request sources including
+// headers, cookies, query parameters, form data, and URL parameters.
+//
+// Key features:
```
Comment (gaby) [inline extractors/extractors.go]: These key features section can be removed. That's what the docs are for

File: extractors/README.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+- `FromParam(param string)`: Extract from URL path parameters
+- `FromForm(param string)`: Extract from form data
+- `FromHeader(header string)`: Extract from custom HTTP headers
+- `FromQuery(param string)`: Extract from URL query parameters
+- `FromCustom(key string, fn func(fiber.Ctx) (string, error))`: Define custom extraction logic with metadata
+- `Chain(extractors ...Extractor)`: Chain multiple extractors with fallback
+
+### Source Inspection
+
+The `Source` field enables security-aware extraction by identifying the origin of extracted values:
```
Comment (gaby) [inline extractors/README.md]: This comment is confusing

File: extractors/README.md [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+- If all extractors fail, returns the last error encountered or `ErrNotFound`
+- **Skips extractors with `nil` Extract functions** (graceful error handling)
+- Preserves metadata from first extractor for introspection
+- Stores defensive copy for runtime inspection via the `Chain` field
+
+## Security Considerations
+
+### Source Awareness and Custom Extractors
+
+The `Source` field enables security-aware extraction by identifying the origin of extracted values. However, when using `FromCustom`, middleware cannot determine the source of the extracted value, which can compromise security:
```
Comment (gaby) [inline extractors/README.md]: How is specifying a field, a security feature?

File: extractors/README.md [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+- **CSRF Protection**: The double-submit-cookie pattern requires tokens to come from cookies. Custom extractors may read from insecure sources without middleware being able to detect or prevent this
+- **Authentication**: Security middleware may not be able to enforce source-specific security policies
+- **Audit Trails**: Source information is lost, making security analysis more difficult
+
+Documentation and examples should clearly warn about these risks when using custom extractors.
+
+## Testing
```
Comment (gaby) [inline extractors/README.md]: This section can be removed. Main readme.md covers testing

File: extractors/README.md [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+- Individual extractor functionality across all source types
+- Error handling and edge cases (empty values, malformed headers)
+- Chained extractor behavior and error propagation
+- Custom extractor support including nil function handling
+- RFC 7235 compliance for Authorization header parsing
+- Metadata validation and source introspection
+- Chain ordering and fallback logic (comprehensive test functions)
+
+## Maintenance Notes
```
Comment (gaby) [inline extractors/README.md]: This section can be removed too

File: docs/guide/extractors.md [hunk 17]
```diff
@@ -0,0 +1,401 @@
+---
+id: extractors
+title: 🔬 Extractors
+description: Understanding how middleware extracts values from HTTP requests
```
Comment (gaby) [inline docs/guide/extractors.md]: I believe this line was from when extractors was intended to be its own thing. It should be refactor

File: docs/guide/extractors.md [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+sidebar_position: 8.5
+toc_max_heading_level: 4
+---
+
+The extractors package provides shared value extraction utilities for Fiber middleware packages. It helps reduce code duplication across middleware packages while ensuring consistent behavior and security practices.
```
Comment (gaby) [inline docs/guide/extractors.md:9]: Same here

File: docs/guide/extractors.md [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+
+## Overview
+
+The `github.com/gofiber/fiber/v3/extractors` package serves as a centralized location for value extraction logic that is common across multiple middleware packages. This approach:
```
Comment (gaby) [inline docs/guide/extractors.md]: Its not a package though

File: docs/guide/extractors.md [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+
+- **Reduces Code Duplication**: Eliminates redundant extractor implementations across middleware packages
+- **Ensures Consistency**: Maintains identical behavior and security practices across all extractors
+- **Simplifies Maintenance**: Changes to extraction logic only need to be made in one place
+- **Enables Direct Usage**: Middleware can import and use extractors directly
+- **Improves Performance**: Shared, optimized extraction functions reduce overhead
+
+## Installation
```
Comment (gaby) [inline docs/guide/extractors.md]: Remove installation section

**Review by gaby — CHANGES_REQUESTED** [review]
The docs/guide/extractors.md needs to be refactor. It's assuming that `extractors` is a separate package, which it isn't anymore

File: extractors/extractors.go (hunk 11 above)
Comment (sixcolors) [reply extractors/extractors.go:101]: I was going to change it in keyauth. FromAuthHeader already implies its Authorization. So having two parameters, while allowing flexibility, didn’t make sense.

File: extractors/README.md (hunk 14 above)
Comment (sixcolors) [reply extractors/README.md]: It’s an introspection feature which middleware like CSRF use for security. Do you have a better wording idea?

**Comment by ReneWerner87** [comment]
@gaby can you check again
